Add group volume feature and related tests to Sonos

TheMegamind opened the pull request:
## Proposed change
This PR adds support for Sonos’ native group volume controls. It adds a new number entity (slider) that exposes each group’s overall volume, as managed by the group coordinator for each group present in the system. 

Aside from viewing the current group volume, the entity (slider) allows users to adjust the overall loudness of the entire group without altering the relative balance of its individual members, which is useful for households or commercial setups where relative speaker balance is tuned but overall loudness needs quick adjustment (e.g., lowering during a phone call, raising for a party).

The group volume entity is available on all players. When a player is grouped, the entity mirrors the volume of its group; when ungrouped, it reflects the individual player’s volume.

**Implementation notes**
•	Fully event-driven; no background polling.
•	Updates are fanned out in process, avoiding redundant queries to every member.
•	Handles volume and topology changes responsively, even in larger systems
•	Includes new tests in test_number.py 


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

TheMegamind committed:
Sonos: Add Group Volume feature and related tests

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+
+        self._unsub_coord: Callable[[], None] | None = None
+        self._unsub_activity: Callable[[], None] | None = None
+        self._unsub_stop: Callable[[], None] | None = None
```
Unneeded?

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+    def _current_group_uid(self) -> str:
+        return self.soco.group.uid
+
+    def _is_grouped(self) -> bool:
+        """Return True if the player is in a group with 2+ members."""
+        group = getattr(self.soco, "group", None)
+        if group is None:
+            return False
+        members = getattr(group, "members", None)
+        return bool(members and len(members) > 1)
+
```
```suggestion
    def _is_grouped(self) -> bool:
        """Return True if the player is in a group with 2 or more members."""
        if (group := getattr(self.soco, "group", None)) and (members := getattr(group, "members", None)):
            return len(members) > 1
        return False
```

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+        members = getattr(group, "members", None)
+        return bool(members and len(members) > 1)
+
+    def _coordinator_soco(self):
+        # member or coordinator: this returns the coordinator SoCo
+        return (self.speaker.coordinator or self.speaker).soco
+
```
```suggestion
    def _coordinator_soco(self) -> SoCo:
        """Return the coordinator SoCo for this speaker."""
        return (self.speaker.coordinator or self.speaker).soco
```

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+                )
+                return None
+
+        vol = await self.hass.async_add_executor_job(_get)
+        if vol is None:
+            return
```
```suggestion
    if (vol := await self.hass.async_add_executor_job(_get)) is None:
        return
```

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+                vol,
+            )
+
+    # ------------------ wiring & lifecycle ----------------
```
remove - not a widespread HA style convention

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+        self._coord_uid = (self.speaker.coordinator or self.speaker).uid
+        self._group_uid = self._current_group_uid()
+
+        # Any activity from any speaker → check if topology changed, refresh once
```
A consideration. 

Whenever group volume is changed the speaker sends a RenderingControl event which is received by speaker.py:async_update_volume.

[speaker.py](https://github.com/themegamind/core/blob/70008c191ac8e50bd53db4feb1b48407d196b564/homeassistant/components/sonos/speaker.py#L566-L568)

When the volume changes for a speaker; send an event with the coordinator uid (e.g.  SONOS_VOLUME_UPDATE-{coord_uid}) and have each number entity subscribed only to this event and then retrieve the group volume from the coordinator soco. This may fix the tracking issue reported and reduce activity fetching group volumes whenever there is unrelated activity.

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+        if self._unsub_activity is not None:
+            self._unsub_activity()
+            self._unsub_activity = None
+        if self._unsub_stop is not None:
+            self._unsub_stop()
+            self._unsub_stop = None
+
```
unneeded

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +164,240 @@ def native_value(self) -> float:
+
+    # ------------------ signal handlers ------------------
+
+    def _bind_coordinator_state(self, coord_uid: str) -> None:
```
_async_bind_coordinator_state

TheMegamind committed:
Reduce overhead and optimize group vol logic for stability and accuracy

TheMegamind committed:
Fix thread-safe scheduling for group volume refresh

TheMegamind commented:
I just pushed new code, as you surely know already This code was on its third day of soak on my own system, both with and without the two dozen or more lines of debug code I added to ensure I had it right this time. It was working great.   

But then it failed one of the CI tests and my rushed attempt to fix it only made things worse.  I'll need extra time to figure it out, as some of this is new to me. 

In the meantime, here's a short summary of the changes in the larger commit (although I understand you'll likely wait to review anything until I fix what's broken). 

* Minimize overhead and churn 
  - eliminate unnecessary rebinds and other noise
  - schedule all dispatcher calls through the event loop for thread-safety
  - improve handling of topology settling delays
  - make the coordinator authoritative for group volume
* Added and revised tests
* Addressed other issues/comments from previous commit 
* Hopefully, you'll find the code to be cleaner as well...

PeteRager commented:
Sounds good,  I'm limited on time next two weeks, may be some delays.

TheMegamind committed:
Fix Sonos group volume rebind/refresh path to unbreak tests

TheMegamind commented on homeassistant/components/sonos/media_player.py:
This change inadvertently hitched a ride with Sunday's push, although I had intended to mention it eventually. It's a quick & dirty fix for an `AssertionError` (resulting from a transient `None` in `group_members`) that showed up occasionally in testing. 

Without the guard, this test (which I wrote to replicate the error) eventually fails, at least on my box.  With the guard, it never fails.  

```
for i in $(seq 1 25); do
  echo "🔁 Run $i/25  $(date -Ins)"
  stdbuf -oL -eL pytest -q tests/components/sonos/test_media_player.py::test_entity_basic \
    -n auto --timeout=30 --maxfail=1 || { echo "❌ Failed on run $i"; break; }
done
```

As this change is beyond the scope of the PR, I assume I should revert it, Just thought I should explain first, since it's already out there.

TheMegamind committed:
Revert sonos media_player.py to upstream/dev to remove unrelated changes

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,408 @@ def native_value(self) -> float:
+        )
+        self.async_on_remove(self._unsubscribe_activity)
+
+        # Coordinator state updates
+        self._unsubscribe_coord = async_dispatcher_connect(
+            self.hass,
+            f"{SONOS_STATE_UPDATED}-{self._coord_uid}",
+            self._on_coord_state_updated,
+        )
+        self.async_on_remove(self._unsubscribe_coord)
+
+        # This member's own updates
+        self._unsubscribe_member = async_dispatcher_connect(
+            self.hass,
+            f"{SONOS_STATE_UPDATED}-{self.speaker.uid}",
+            self._on_member_state_updated,
+        )
+        self.async_on_remove(self._unsubscribe_member)
+
+        # Subscribe to current group fan-out (if any)
+        self._subscribe_group_fanout(self._group_uid)
+
+        # Coordinator listens for refresh requests for its group (if applicable)
+        self._subscribe_group_requests_if_coord(self._group_uid)
+
+        # Kick an initial rebind + populate, then a delayed settle read
+        self._rebind_for_topology_change()
+        await self._async_initial_populate()
```
This is similar to the code in _rebind_for_topology_change. Can that be used instead?

TheMegamind replied:
Ooh, I should have caught that. 🤦🏻‍♂️

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,408 @@ def native_value(self) -> float:
+        self._rebind_for_topology_change()
+
+    @callback
+    def _on_group_volume_fanned(self, payload: tuple[str, int]) -> None:
```
I like this, great way to fan out the volume and not having to refresh it on each entity of the group

TheMegamind commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,408 @@ def native_value(self) -> float:
+
+            vol = await self.hass.async_add_executor_job(_get_player)
+
+        if vol is not None and self._value != vol:
```
This was done intentionally.  Transient read failures from SoCo (timeouts, socket closed) are not uncommon, and if we set set self._value = None on every failure, the group slider flickers to `unknown` and back with every tiny blip in connectivity. Doing it this way seemed cleaner, given that the `available` property already reflects whether the speaker is online.

TheMegamind committed:
B-6

TheMegamind committed:
B-7

TheMegamind committed:
Incorporate reviewer-requested changes to number.py and test_number.py

TheMegamind committed:
Incorporate reviewer-requested changes to number.py and test_number.py

TheMegamind commented:
The latest commit incorporates the changes you requested. I hope I'm getting closer with the tests, which I think are testing me more than the feature addition at this point.  

In that regard, please note that `test_coordinator_listener_rebinds_on_coord_change` has been temporarily removed, as it has me stymied, I'm going to work on it separately as time allows.  Thanks.

PeteRager commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,369 @@ def native_value(self) -> float:
+                self._rebind_for_topology_change()
+
+                # During bootstrap, let members populate from the coordinator once
+                if (
+                    self._bootstrap
+                    and self._is_grouped()
+                    and not self._is_coordinator()
+                    and self._value is None
+                ):
+                    await self._async_initial_populate()
```
The tests do not cover this section of the code. HA does not execute line 254 when running in debug mode; starting up with speakers that are grouped or un-grouped.

TheMegamind commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,369 @@ def native_value(self) -> float:
+    """
+
+    _attr_translation_key = "group_volume"
+    _attr_icon = "mdi:volume-high"
```
Nope, that's on me.  I should have reviewed for understanding before making that change.

TheMegamind committed:
Fix icon, remove unused code, revise tests & add new test

TheMegamind commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,372 @@ async def test_amp_number_entities(
+
+
+# Subclass to expose and test internal rebinding logic
+class _TestEntity(SonosGroupVolumeEntity):
```
You beat me to the punch. A colleague I asked to look over things told me earlier today that the method was suspect. When I explained the trouble I'd been having, he asked me a simple question:  Had my previous efforts had accounted for the refresh_delay? You could almost hear the click as the light bulb over my head turned on. 

There's a good chance I'll have a solution for you this evening. If not, I'll post test_number.py without that test, as requested.

TheMegamind committed:
Replace rebinds test with conforming version

TheMegamind commented:
Submitted with new rebind test and no other changes.  I tried closing and reopening the CI to rerun the tests when they failed the first, but I'll have to leave it there for tonight.

PeteRager commented:
> Submitted with new rebind test and no other changes. I tried closing and reopening the CI to rerun the tests when they failed the first time, but I'll have to leave it there for tonight.

You can hit update branch and that'll rerun the CI or push an empty commit.  

```
git commit --allow-empty -m "ci: retrigger workflow"
git push
```

TheMegamind committed:
Merge branch 'dev' into feature/sonos_group_volume_new

TheMegamind committed:
Merge branch 'dev' into feature/sonos_group_volume_new

PeteRager reviewed (commented):
A couple of small changes. And awaiting the future of that test case.

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,377 @@ async def test_amp_number_entities(
+    # Grab the group-volume entity object created by the fixture
+    entity = next(
+        ent
+        for ent in hass.data["entity_components"]["number"].entities
```
This is also considered accessing internals.

TheMegamind commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,377 @@ async def test_amp_number_entities(
+    # State updates on activity (refresh is scheduled -> advance time)
+    async_dispatcher_send(hass, SONOS_SPEAKER_ACTIVITY, "test")
+    await hass.async_block_till_done()
+    async_fire_time_changed(hass, dt_util.utcnow() + timedelta(seconds=1))
```
Yes.  Did that this morning, but was awaiting any other feedback you might have.  

```
from homeassistant.components.sonos.number import GV_REFRESH_DELAY
...
# Advance time so delayed refresh runs
async_fire_time_changed(hass, utcnow() + timedelta(seconds=GV_REFRESH_DELAY + 0.1))
await hass.async_block_till_done()
`

TheMegamind commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,377 @@ async def test_amp_number_entities(
+    """When group uid changes, entity rebinds; polling refresh reflects new group volume."""
+    # Start grouped with deterministic G-1 and bring up the platform
+    gid1 = _force_grouped(soco)
+    if gid1 != "G-1":
```
Ugh, a bad habit I need to break.  But an easy fix.

TheMegamind committed:
Additional tweaks to tests

TheMegamind committed:
Merge branch 'feature/sonos_group_volume_new' of https://github.com/TheMegamind/core into feature/sonos_group_volume_new

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+        self.soco = MagicMock()
+
+
+class _FakeSpeaker:
+    def __init__(self, uid: str, coord_uid: str, group_uid="G-1") -> None:
+        # Represent a SonosSpeaker with soco + coordinator attributes
+        self.uid = uid
+        self.zone_name = f"Zone-{uid}"
+        self.soco = _FakeSoCo(
+            uid=f"SOCO-{uid}",
+            zone_name=self.zone_name,
+            group_uid=group_uid,
+            coord_uid=coord_uid,
+        )
+        self.coordinator = _FakeCoordinator(coord_uid)
+        self.available = True
```
This appears to not be used anymore?

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+    assert state.state == STATE_UNKNOWN
+
+
+class _FakeSoCoGroup:
+    def __init__(self, uid: str, coordinator=None, members=None) -> None:
+        # Represent a SoCo group with uid, coordinator, and members
+        self.uid = uid
+        self.coordinator = coordinator
+        self.members = members or ["m1", "m2"]
+
+
+class _FakeSoCo:
+    def __init__(
+        self, uid: str, zone_name="Z1", group_uid="G-1", coord_uid="C-OLD"
+    ) -> None:
+        # Represent a SoCo device with its current group and coordinator
+        self.uid = uid
+        self.zone_name = zone_name
+        coord = MagicMock()
+        coord.uid = coord_uid
+        self.group = _FakeSoCoGroup(group_uid, coordinator=coord)
+
+
+class _FakeCoordinator:
+    def __init__(self, uid: str) -> None:
+        # Represent a speaker coordinator with a soco handle
+        self.uid = uid
+        self.soco = MagicMock()
```
These classes do not appear to be used anymore?

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+
+
+async def test_group_volume_rebinds_on_topology_change(
+    hass: HomeAssistant, async_setup_sonos, soco
```
add typing to soco.

soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+
+
+async def test_group_volume_exception(
+    hass: HomeAssistant, async_setup_sonos, soco
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+async def test_group_fanout_unsubscribe_and_resubscribe_on_group_change(
+    hass: HomeAssistant,
+    async_setup_sonos,
+    soco,
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+
+
+async def test_group_volume_ungrouped_sets_player_volume(
+    hass: HomeAssistant, async_setup_sonos, soco
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+async def test_group_volume_updates_on_activity(
+    hass: HomeAssistant,
+    async_setup_sonos,
+    soco,
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+
+
+async def test_group_volume_sets_backend_and_updates_state(
+    hass: HomeAssistant, async_setup_sonos, soco
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+        await async_setup_sonos()
+
+
+def _force_grouped(soco) -> str:
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+
+
+async def test_group_volume_rounds_in_range(
+    hass: HomeAssistant, async_setup_sonos, soco
```
typing for soco parameter soco: MockSoCo

PeteRager commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +109,326 @@ async def test_amp_number_entities(
+async def test_group_volume_fallback_polling(
+    hass: HomeAssistant,
+    async_setup_sonos,
+    soco,
```
typing for soco parameter soco: MockSoCo

PeteRager reviewed (changes requested):
Couple of typing missed and some unused code (I think); nice work on the updated test case!

TheMegamind committed:
Remove used code. Add missing typing.

TheMegamind commented:
> We are all volunteers - and we do the best we can when we can.

My appreciation for all the volunteers do is without limit—especially when you're working with people like me, who know just enough about coding to be dangerous. (And who sometimes create unnecessary work, like neglecting to remove unneeded code 🤦🏻‍♂️). I hope my comments didn't imply anything else.  

The PR has been much more involved than I expected, but I've learned a lot. Thanks for all your effort.

TheMegamind committed:
Revise parameter typing.

TheMegamind committed:
fix import order/formatting in test_number.py (ruff)

TheMegamind committed:
fix import order/formatting in test_number.py (ruff)

PeteRager reviewed (approved):
Nice work. Thank you.

TheMegamind commented:
> Nice work. Thank you.

Unsure if I missed a step.  Is there anything I need to do to help this move forward? Just checking.  🤷🏻‍♂️

emontnemery committed:
Merge branch 'dev' into feature/sonos_group_volume_new

emontnemery commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+                    _gv_req_signal(group_uid),
+                    None,
+                )
+                self._schedule_delayed_refresh(GV_REFRESH_DELAY)
```
```suggestion
                self._schedule_delayed_refresh(GROUP_VOLUME_REFRESH_DELAY)
```

emontnemery commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+        """Coordinator-only: handle a member’s request to refresh group volume."""
+        if not (self._is_grouped() and self._is_coordinator()):
+            return
+        self._schedule_delayed_refresh(GV_REFRESH_DELAY)
```
```suggestion
        self._schedule_delayed_refresh(GROUP_VOLUME_REFRESH_DELAY)
```

emontnemery commented on tests/components/sonos/test_number.py:
```diff
@@ -94,3 +117,284 @@ async def test_amp_number_entities(
+    async_dispatcher_send(hass, SONOS_SPEAKER_ACTIVITY, "test")
+    await hass.async_block_till_done()
+    async_fire_time_changed(
+        hass, dt_util.utcnow() + timedelta(seconds=GV_REFRESH_DELAY + 0.1)
```
```suggestion
        hass, dt_util.utcnow() + timedelta(seconds=GROUP_VOLUME_REFRESH_DELAY + 0.1)
```

emontnemery commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+        self._unsubscribe_member: Callable[[], None] | None = None
+        self._unsubscribe_activity: Callable[[], None] | None = None
+        self._unsubscribe_gv_signal: Callable[[], None] | None = None
+        self._unsubscribe_gv_req: Callable[[], None] | None = None
```
Same comment as for `_unsubscribe_gv_signal`

emontnemery commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+        self._delay_unsubscribe: Callable[[], None] | None = None
+
+        self._value: int | None = None
+        self._last_rebind_time: float = 0.0
```
Make it clear from the variable name it's a timstamp; end it with `_ts` or `_timestamp`, not `_time`

emontnemery commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+        """One-time populate before the coordinator fans out values."""
+        if self._is_grouped():
+            # Read the coordinator's group volume directly
+            def _get_group() -> int | None:
```
```suggestion
            def _get_group_volume() -> int | None:
```

emontnemery commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+                    )
+                    return None
+
+            vol = await self.hass.async_add_executor_job(_get_player)
```
These are duplicated in `_async_refresh_from_device`, please deduplicate

TheMegamind replied:
Ooh, I should have caught that myself. Thanks.

TheMegamind committed:
Apply suggestions from code review

Co-authored-by: Erik Montnemery <erik@montnemery.com>

TheMegamind committed:
Additional revisions per reviewer feedback

TheMegamind committed:
Apply Prettier formatting

TheMegamind commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+
+            self._delay_unsubscribe = async_call_later(
+                self.hass, seconds, _delayed_refresh
+            )
```
I experimented with using the Debouncer helper originally but eventually settled on the more explicit scheduling logic here, since it handles a few things the helper doesn’t — notably _rebind_for_topology_change(), loop-state checks, and explicit cancellation during entity unload.

The goal was to ensure only one topology rebind and refresh occurs even on larger systems where group topology and volume changes can happen rapidly, given all the inter-entity dependencies involved. The current _schedule() logic has handled those cases reliably in weeks of live testing.

Of course, it *might* be possible to wrap Debouncer to cover these needs but personally I think that adds complexity without any real practical benefit.

TheMegamind commented on homeassistant/components/sonos/number.py:
```diff
@@ -142,3 +173,357 @@ def native_value(self) -> float:
+            members := getattr(group, "members", None)
+        ):
+            return len(members) > 1
+        return False
```
We use `getattr()` here defensively because `SoCo.group` and its nested attributes (`uid`, `members`, etc.) can be briefly unset or invalidated during group-topology changes..

Direct access (e.g, `self._coordinator_soco().group.uid`) can raise `AttributeError` during those transient states. Using `getattr(..., default=None)` avoids those race-condition errors and lets `_rebind_for_topology_change()` handle regrouping cleanly when the topology stabilizes.

TheMegamind commented:
I’m going to pause this PR while I consider the best path forward. At this point my dev branch has gone stale, and while I could try to rebase it, I haven’t had the best of luck rebasing long-lived PR branches in the past.

Separately, I’m seeing runtime warnings around blocking I/O in the event loop that I don’t believe were there before. I haven’t determined whether that’s due to recent Core changes, stricter runtime checks, or something else entirely, but either way the wisest course may be to re-examine the approach.

While I still very much prefer the fan-out approach in this PR, my current thinking is to reopen this as a smaller, coordinator-only group volume change on a clean baseline. My thinking is that this would be less fragile and more likely to move through review in a reasonable timeframe.
